src/librustc_trans/base.rs

```diff
@@ -1145,6 +1145,23 @@ pub fn trans_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     };
     let no_builtins = attr::contains_name(&krate.attrs, "no_builtins");
 
+    // Skip crate items and just output metadata in -Z no-trans mode.
+    if tcx.sess.opts.debugging_opts.no_trans ||
+       tcx.sess.opts.output_types.contains_key(&config::OutputType::Metadata) {
```

Comment:
I just so that there's already `OutputTypes::should_trans()` which does what's needed here. I'll change it.